### PR TITLE
[AMDGPU] RFC: Add and optionally use GCNIterativeRPTrackers

### PR DESCRIPTION
**Pull request by jrbyrnes**
Having walked through several interesting lit changes, I think it's a good time to put up a WIP PR for any high level comments.

There are still a few things left to do, namely:

1. 	Verify more lit changes when enabled
2. 	Add tests
3. 	Compatibility with schedulers in GCNIterativeScheduler.cpp 

Nonetheless, this patch adds the GCNIterativeRPTrackers. We can optionally use these instead of generic RPTrackers during scheduling. The benefits include: 1. better subreg liveness modeling, 2. actually tracking %av registers (counted as VGPRs), and 3. having accurate global live-through RP.

I've introduced as separate classes since there is other outstanding work on RPTrackers. The plan is to merge the two projects once at steady state.

### CODE REVIEW

**Comment by vpykhtin** [comment]
Hi @jrbyrnes, would be nice if you add a few words explaining overall approach. I'm curious why these new trackers are iterative, is it just a name to distinguish them from current?

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 1]
```diff
@@ -343,6 +343,48 @@ void GCNRPTracker::reset(const MachineInstr &MI,
   MaxPressure = CurPressure = getRegPressure(*MRI, LiveRegs);
 }
 
+DenseMap<int, GCNRPTracker::LiveRegSet>
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:346]: I'm not sure what the key is here. I'm also concerned by building maps for liveness info, that's already encoded in the LiveIntervals. Most of this code is also replicating logic from the generic tracker, and all of this is really hard to get correct.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp [hunk 2]
```diff
@@ -570,6 +612,127 @@ bool GCNUpwardRPTracker::isValid() const {
...
+
+  // Remove dead registers or mask bits.
+  SmallSet<Register, 8> SeenRegs;
+  for (auto &MO : MI.operands()) {
+    if (!MO.isReg() || !MO.getReg().isVirtual())
+      continue;
+    if (!MO.isUse())
+      continue;
+    if (!MO.readsReg())
+      continue;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:702]: Checking !isUse before !readsReg is almost certainly wrong. You almost always want readsReg over isUse

**Comment by jrbyrnes** [comment]
> Would be nice if you add a few words explaining overall approach. I'm curious why these new trackers are iterative, is it just a name to distinguish them from current?

I'll explain the approach by defining the issues. 

`GCNSchedStrategy` uses `GCNDownwardRPTracker` in a couple different places to calculate the pressure of a region. At these points, the schedule is in a steady state, so the tracker can scan through the region in order. Thus, it is okay to internalize the iterators (`LastTrackedMI`, `NextMI`, `MBBEnd`), which the tracker maintains through calls to `advance()` / `advanceBeforeNext()` / `advanceToNext()`. Obviously during scheduling the order is subject to change, so these internal iterators don't make sense -- at least not in the way they are currently maintained. Moreover, with these, there is no good way to query the impact a candidate `SU` will have on RP (`initCandidate`) since calls to `advance*` will update the iterators.

On the other hand, `GCNIterativeDownwardRPTracker` does not assume an in-order scan of a region, so it does not suffer from the above problems. I named it iterative because we are incrementally constructing the order, but I'm not very confident about the naming. Note that the `GCNUpwardRPTracker` has a more friendly interface for this use-case, but I also implemented `GCNIterativeUpwardRPTracker` since scheduling is bi-directional, and having a shared base class is the cleaner approach.

As noted in the code, the actual RP calculation implemention is largely copy & paste from the existing trackers. Note: when calculating RP, we call `getRegKind` which treats `%AV` registers as `VGPR`. During scheduling, we use the trackers to update the state of RP when we `schedNode`, and use this state to measure the impact of candidate `SU` in `initCandidate`. When first scheduling a region, we initialize the tackers with the LiveIn (Downward) / LiveOut (Upward) registers as calculated by the existing `getLiveRegMap`. This calculation is global live-thru aware (as opposed to the generic tracker which isn't https://github.com/llvm/llvm-project/blob/047b2b241defcad79a6ac0fec9cda092bac0a922/llvm/lib/CodeGen/RegisterPressure.cpp#L355 ).

The generic trackers are still maintained through scheduling, but the meaningful usecase of them is when picking between candidate SUs, and that usecase has been replaced with our trackers (`initCandidate`).

**Comment by vpykhtin** [comment]
Ok, thanks. Just on a historical note upward and downward trackers were introduced for different purposes. Upward tracker was designed with two ideas in mind:

1. To be able tracking unimplemented schedule, that is the order if instruction is maintained externally, for example in array.
2. The LIS isn't updated for the unimplemented schedule.

In this sense it is iterative because it was supposed to use on iterations over different schedules.

Currently I have the need to be able to collect live registers at the point of maximum pressure over an instruction. It seems that it's not the same set that would be collected iterating over slot index R because an instruction define and use registers at R slot index and we would consider 'dest' and 'source' registers as live at COPY instruction, the mistake that upward tracker had previously.

For that reason I would add some interface to be able to collect live registers at different stages of instruction execution.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp (hunk 1 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp]: I'm not sure what to do with this comment. 

> I'm not sure what the key is here.

Currently, these maps are keyed on the initial first instructions of the regions. However, if we are going to requery, we need a different key as these first instructions are likely to change. I've used instead RegionIdx -- which tracks the index in the Regions array. This is similarly used to index all these https://github.com/llvm/llvm-project/blob/a5044e6d505deb79f1b00bb39d11096d29b9c910/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp#L623C3-L623C10 I guess we could restructure as vector.

> I'm concerned by building maps for liveness info, that's already encoded in the LiveIntervals. 

While this patch does introduce a LiveOutMap, it doesn't introduce the notion of using a Map to cache live-ins -- this already exists in the code. In this patch, these maps optimize compile time by avoiding recalculation of the live regs. 

> Most of this code is also replicating logic from the generic tracker, and all of this is really hard to get correct.

This is mostly copy-paste from existing code in GCNRegPressure, but, nonetheless, we're running a CQE cycle to root out any correctness / performance issues. Per discussion, I think it's better to slightly duplicate logic for time being, and potentially rework the generic tracker in a long term project.

**Comment by jrbyrnes** [comment]
Ping for any other high level comments while I transition this to ready-for-review?

**Comment by vpykhtin** [comment]
Sorry but this is very hard to review. For example, you've moved getLiveRegMap to the new location and modified it at the same time.  Also, I start to think it wasn't a good idea to duplicate tracker's code because I cannot see the diffs. 

If I understood correctly, you can reuse gcn upward tracker in it's current state? You only need to introduce iterative downward tracker?

**Comment by jrbyrnes** [comment]
I plan to split up the PR and refactor a bit to make it easier on the reviewer.

File: llvm/lib/Target/AMDGPU/GCNRegPressure.cpp (hunk 1 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/GCNRegPressure.cpp]: > I'm not sure what the key is here. I'm also concerned by building maps for liveness info, that's already encoded in the LiveIntervals. 

I've tried to address this in https://github.com/llvm/llvm-project/pull/93089

I've left getLiveRegMap unmodified and used the last MI per region as the key to the liveness info (as is similarly done for BBLiveInMap currently). The result is that we do not introduce any new way of building maps for liveness info. The maps are not strictly necessary, and we could recompute them each time we encounter the same region during scheduling, but they do eliminate redundant calculations.

**Comment by jrbyrnes** [comment]
Closing in favor of the stack of PRs:

Create BBLiveOutMap using existing getLiveRegMap: https://github.com/llvm/llvm-project/pull/93089
Extend Current Trackers for External iterator: https://github.com/llvm/llvm-project/pull/93088
Scheduler changes necessary to use new trackers: https://github.com/llvm/llvm-project/pull/93090

**Pull request closed without merging**